app/assets/javascripts/oxalis/view/right-menu/tree_hierarchy_view.js

```diff
@@ -0,0 +1,481 @@
+        _.orderBy(groupToTreesMap[groupId], [sortBy], ["asc"]).map(makeExtendedGroupObjectFromTree),
+      );
+    }
+    transformedGroup.isChecked = _.every(
+      transformedGroup.children,
+      groupOrTree => groupOrTree.isChecked,
+    );
+    return transformedGroup;
+  });
+}
+
+function callDeep(
+  groups: Array<TreeGroupType>,
+  groupId: number,
+  callback: (TreeGroupType, number, Array<TreeGroupType>, ?number) => void,
+  parentGroupId: ?number = MISSING_GROUP_ID,
+) {
+  groups.forEach((group: TreeGroupType, index: number, array: Array<TreeGroupType>) => {
+    if (group.groupId === groupId) {
+      callback(group, index, array, parentGroupId);
+    }
+    if (group.children) {
+      callDeep(group.children, groupId, callback, group.groupId);
+    }
+  });
+}
+
+function createGroupToTreesMap(trees: TreeMapType): { [number]: Array<TreeType> } {
+  return _.groupBy(trees, tree => (tree.groupId != null ? tree.groupId : MISSING_GROUP_ID));
+}
+
+class TreeHierarchyView extends React.PureComponent<Props, State> {
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+    if (
+      prevState.prevProps == null ||
+      prevState.prevProps.trees !== nextProps.trees ||
+      prevState.prevProps.treeGroups !== nextProps.treeGroups ||
+      prevState.prevProps.sortBy !== nextProps.sortBy
+    ) {
+      // Insert the trees into the corresponding groups and create a
+      // groupTree object that can be rendered using a SortableTree component
+      const groupToTreesMap = createGroupToTreesMap(nextProps.trees);
+      const rootGroup = {
+        name: "Root",
+        groupId: MISSING_GROUP_ID,
+        children: nextProps.treeGroups,
+      };
+      const expandedGroupIds = _.cloneDeep(prevState.expandedGroupIds);
+      const generatedGroupTree = insertTreesAndTransform(
+        [rootGroup],
+        groupToTreesMap,
+        expandedGroupIds,
+        nextProps.sortBy,
+      );
+      return {
+        groupTree: generatedGroupTree,
+        expandedGroupIds,
+        prevProps: nextProps,
+      };
+    } else {
+      return {
+        prevProps: nextProps,
+      };
+    }
+  }
+
+  state = {
+    hoveredGroupId: null,
+    expandedGroupIds: {},
+    groupTree: [],
+    renamingGroup: null,
+    // TODO: Remove once https://github.com/yannickcr/eslint-plugin-react/issues/1751 is merged
+    // eslint-disable-next-line react/no-unused-state
+    prevProps: null,
+  };
+
+  onChange = treeData => {
+    this.setState({ groupTree: treeData });
+  };
+
+  onCheck = evt => {
+    const { groupId, type } = evt.target.node;
+    if (type === TYPE_TREE) {
+      this.props.onToggleTree(parseInt(groupId, 10));
+    } else if (groupId === MISSING_GROUP_ID) {
+      this.props.onToggleAllTrees();
+    } else {
+      this.props.onToggleTreeGroup(groupId);
+    }
+  };
+
+  onSelect = evt => {
+    const treeId = evt.target.dataset.groupid;
```

Comment:
Every "node" in the tree hierarchy has a groupId, for nodes with type `TYPE_GROUP` this is the groupId and for nodes with type `TYPE_TREE` this is the treeId. I can follow your confusion though, I did it this way because that way I could simply extend the `TreeGroupType` to be used in the tree hierarchy data structure. Maybe this should be rewritten to be more clear.

Reply:
Maybe it would help to rename `groupId` to `id` everywhere where it is used within a tree group object (additionally, that would avoid expressions like `group.groupId` in favor of `group.id`)? I'm not familiar with the code, though. So, maybe it doesn't help ^^

Reply:
I decoupled the ExtendedTreeGroupType from the TreeGroupType. The ExtendedTreeGroupType is now (more descriptively imo) named TreeNodeType as it represents the tree nodes that are rendered by the react-sortable-tree (which are either groups or skeleton trees). I used this opportunity to rename groupId to id for these tree nodes, as you proposed :)